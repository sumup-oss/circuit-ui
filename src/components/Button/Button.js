/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { omit } from 'lodash/fp';
import PropTypes from 'prop-types';

import PlainButton from './components/PlainButton';
import RegularButton from './components/RegularButton';
import { sizes } from '../../styles/constants';

const { KILO, MEGA, GIGA } = sizes;

const SHARED_PROPS = [
  'children',
  'className',
  'data-selector',
  'disabled',
  'href',
  'onClick',
  'primary',
  'size',
  'type'
];

export const BUTTON_PROPS = [
  ...SHARED_PROPS,
  'deepRef',
  'as',
  'secondary',
  'size',
  'stretch',
  'target'
];

export const PLAIN_BUTTON_PROPS = [...SHARED_PROPS, 'size', 'target'];

const REGULAR_BUTTON_ONLY_PROPS = ['deepRef', 'as', 'secondary', 'stretch'];

/**
 * A button component with support for the anchor and button
 * element as well as a button-looking button and a text link.
 */
const Button = ({ plain, ...props }) =>
  plain ? (
    <PlainButton {...omit(REGULAR_BUTTON_ONLY_PROPS, props)} />
  ) : (
    <RegularButton {...props} />
  );

Button.KILO = KILO;
Button.MEGA = MEGA;
Button.GIGA = GIGA;

Button.propTypes = {
  /**
   * Should the Button be disabled?
   */
  disabled: PropTypes.bool,
  /**
   * URL the Button should lead to. Causes the Button to render an <a> tag.
   */
  href: PropTypes.string,
  /**
   * Makes the button look and behave like a text link.
   */
  plain: PropTypes.bool,
  /**
   * Renders a primary button using the brand color.
   */
  primary: PropTypes.bool,
  /**
   * Size of the button. Use the Button's KILO, MEGA, or GIGA properties.
   */
  size: PropTypes.oneOf([Button.KILO, Button.MEGA, Button.GIGA]),
  /**
   * Renders a secondary button.
   */
  secondary: PropTypes.bool,
  /**
   * Trigger stretch (full width) styles on the component.
   */
  stretch: PropTypes.bool,
  /**
   * Link target. Should only be passed, if href is passed, too.
   */
  target: PropTypes.string
};

Button.defaultProps = {
  disabled: false,
  href: null,
  plain: false,
  primary: false,
  secondary: false,
  size: Button.MEGA,
  stretch: false,
  target: null
};

/**
 * @component
 */
export default Button;
