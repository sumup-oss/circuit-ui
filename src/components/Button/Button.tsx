/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { HTMLProps, ReactNode, ReactElement } from 'react';
import { css } from '@emotion/core';

import styled, { StyleProps } from '../../styles/styled';
import {
  textMega,
  disableVisually,
  focusOutline
} from '../../styles/style-helpers';
import { useComponents } from '../ComponentsContext';

export interface BaseProps {
  children: ReactNode;
  /**
   * Choose from 3 style variants. Default: 'primary'.
   */
  variant?: 'primary' | 'secondary' | 'tertiary';
  /**
   * Choose from 3 sizes. Default: 'mega'.
   */
  size?: 'kilo' | 'mega' | 'giga';
  /**
   * Visually and functionally disable the button.
   */
  disabled?: boolean;
  /**
   * Stretch the button across the full width of its parent.
   */
  stretch?: boolean;
}

type LinkElProps = Omit<HTMLProps<HTMLAnchorElement>, 'size' | 'type'>;
type ButtonElProps = Omit<HTMLProps<HTMLButtonElement>, 'size' | 'type'>;

export type ButtonProps = BaseProps & LinkElProps & ButtonElProps;

type ReturnType = ReactElement<any, any> | null;

const BORDER_WIDTH = '1px';

const baseStyles = ({ theme }: StyleProps) => css`
  label: button;
  display: flex;
  align-items: center;
  justify-content: center;
  width: auto;
  height: auto;
  cursor: pointer;
  ${textMega({ theme })};
  text-align: center;
  text-decoration: none;
  font-weight: ${theme.fontWeight.bold};
  border-radius: ${theme.borderRadius.giga};
  border-width: ${BORDER_WIDTH};
  border-style: solid;

  &:focus {
    ${focusOutline({ theme })};
  }

  &:disabled,
  &[disabled] {
    ${disableVisually()};
  }
`;

const primaryStyles = ({
  theme,
  variant = 'primary'
}: ButtonProps & StyleProps) =>
  variant === 'primary' &&
  css`
    label: button--primary;
    background-color: ${theme.colors.p500};
    border-color: ${theme.colors.p500};
    color: ${theme.colors.white};

    &:hover {
      background-color: ${theme.colors.p700};
      border-color: ${theme.colors.p700};
    }

    &:active {
      background-color: ${theme.colors.p900};
      border-color: ${theme.colors.p900};
    }
  `;

const secondaryStyles = ({
  theme,
  variant = 'primary'
}: ButtonProps & StyleProps) =>
  variant === 'secondary' &&
  css`
    label: button--secondary;
    background-color: ${theme.colors.white};
    border-color: ${theme.colors.n500};
    color: ${theme.colors.n800};

    &:hover {
      background-color: ${theme.colors.n100};
      border-color: ${theme.colors.n700};
    }

    &:active {
      background-color: ${theme.colors.n200};
      border-color: ${theme.colors.n700};
    }
  `;

const tertiaryStyles = ({
  theme,
  variant = 'primary'
}: ButtonProps & StyleProps) =>
  variant === 'tertiary' &&
  css`
    label: button--tertiary;
    background-color: transparent;
    border-color: transparent;
    color: ${theme.colors.p500};

    &:hover {
      color: ${theme.colors.p700};
    }

    &:active {
      color: ${theme.colors.p900};
    }
  `;

const sizeStyles = ({ theme, size = 'mega' }: ButtonProps & StyleProps) => {
  const sizeMap = {
    kilo: `${theme.spacings.bit} calc(${theme.spacings.mega} - ${BORDER_WIDTH})`,
    mega: `${theme.spacings.byte} calc(${theme.spacings.giga} - ${BORDER_WIDTH})`,
    giga: `${theme.spacings.kilo} calc(${theme.spacings.tera} - ${BORDER_WIDTH})`
  };

  return css({
    label: `button--${size}`,
    padding: sizeMap[size]
  });
};

const stretchStyles = ({ stretch }: ButtonProps) =>
  stretch &&
  css`
    label: button--stretch;
    width: 100%;
  `;

const BaseButton = styled('button')<ButtonProps>(
  baseStyles,
  primaryStyles,
  secondaryStyles,
  tertiaryStyles,
  sizeStyles,
  stretchStyles
);

/**
 * The Button component enables the user to perform an action or navigate
 * to a different screen.
 */
export function Button(props: BaseProps & LinkElProps): ReturnType;
export function Button(props: BaseProps & ButtonElProps): ReturnType;
export function Button(props: ButtonProps): ReturnType {
  const { Link } = useComponents();
  const LinkButton = BaseButton.withComponent(Link);
  return props.href ? <LinkButton {...props} /> : <BaseButton {...props} />;
}
