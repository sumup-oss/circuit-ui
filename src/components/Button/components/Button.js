import React from 'react';
import styled, { css } from 'react-emotion';
import PropTypes from 'prop-types';

import { textMega } from '../../../styles/style-helpers';
import { sizes } from '../../../styles/constants';

const { KILO, MEGA, GIGA } = sizes;

const calculatePadding = ({ theme, size: buttonSize }) => (diff = '0px') => {
  const sizeMap = {
    [KILO]: `calc(${theme.spacings.bit} - ${diff}) calc(${
      theme.spacings.mega
    } - ${diff})`,
    [MEGA]: `calc(${theme.spacings.byte} - ${diff}) calc(${
      theme.spacings.giga
    } - ${diff})`,
    [GIGA]: `calc(${theme.spacings.kilo} - ${diff}) calc(${
      theme.spacings.tera
    } - ${diff})`
  };

  if (!sizeMap[buttonSize] && buttonSize) {
    return null;
  }

  return sizeMap[buttonSize] || sizeMap.mega;
};

const disabledStyles = css`
  label: button--disabled;
  opacity: 0.4;
  pointer-events: none;
`;

const stretchStyles = ({ stretch }) =>
  stretch &&
  css`
    label: button--stretched;
    width: 100%;
    display: block;
  `;

const baseStyles = ({ theme, href, ...otherProps }) => css`
  label: button;
  background-color: ${theme.colors.n100};
  border-color: ${theme.colors.n300};
  border-radius: ${theme.borderRadius.mega};
  border-style: solid;
  border-width: 1px;
  box-shadow: inset 0 1px 0 1px rgba(255, 255, 255, 0.06);
  display: ${href ? 'inline-block' : 'block'};
  color: ${theme.colors.n700};
  cursor: pointer;
  font-weight: ${theme.fontWeight.bold};
  width: auto;
  height: auto;
  text-decoration: none;
  ${textMega({ theme })};

  &:active {
    background-color: ${theme.colors.n300};
    border-color: ${theme.colors.n500};
    box-shadow: inset 0 4px 8px 0 rgba(12, 15, 20, 0.3);
  }

  &:focus {
    border-color: ${theme.colors.n500};
    border-width: 2px;
    outline: 0;
    padding: ${calculatePadding({ theme, ...otherProps })('1px')};
  }

  &:hover {
    background-color: ${theme.colors.n300};
  }

  &:hover,
  &:active {
    border-color: ${theme.colors.n500};
    border-width: 1px;
    padding: ${calculatePadding({ theme, ...otherProps })()};
  }

  &[disabled],
  &:disabled {
    ${disabledStyles};
  }
`;

const primaryStyles = ({ theme, primary }) =>
  primary &&
  css`
    label: button--primary;
    background-color: ${theme.colors.p500};
    border-color: ${theme.colors.p700};
    color: ${theme.colors.white};

    &:active {
      background-color: ${theme.colors.p700};
      border-color: ${theme.colors.p900};
    }

    &:focus {
      border-color: ${theme.colors.p700};
    }

    &:hover {
      background-color: ${theme.colors.p700};
    }

    &:hover,
    &:active {
      border-color: ${theme.colors.p900};
    }
  `;

const flatStyles = ({ theme, flat, secondary, ...otherProps }) =>
  flat &&
  !secondary &&
  css`
    label: button--flat;
    border-width: 0px;
    box-shadow: 0 0 0 1px rgba(12, 15, 20, 0.02),
      0 2px 2px 0 rgba(12, 15, 20, 0.06), 0 4px 4px 0 rgba(12, 15, 20, 0.06);

    &:active {
      background-color: ${theme.colors.p900};
      box-shadow: 0 0 0 1px rgba(12, 15, 20, 0.02),
        0 0 1px 0 rgba(12, 15, 20, 0.06), 0 2px 2px 0 rgba(12, 15, 20, 0.06);
    }

    &:active:focus,
    &:hover:focus {
      border-width: 0;
      padding: ${calculatePadding({ theme, flat, secondary, ...otherProps })()};
    }

    &:focus {
      border-width: 2px;
      padding: ${calculatePadding({ theme, flat, secondary, ...otherProps })(
        '2px'
      )};
    }
  `;

const secondaryStyles = ({ theme, secondary, flat, ...otherProps }) =>
  secondary &&
  css`
    label: button--secondary;
    background-color: transparent;
    border-color: ${theme.colors.n900};
    border-width: 0;
    box-shadow: none;
    color: ${theme.colors.n700};

    &:active {
      background-color: transparent;
      border-color: ${theme.colors.n900};
      border-width: 0;
      box-shadow: none;
    }

    &:focus {
      border-color: ${theme.colors.n900};
      border-width: 2px;
      box-shadow: none;
      padding: ${calculatePadding({ theme, flat, secondary, ...otherProps })(
        '2px'
      )};
    }

    &:hover {
      background-color: transparent;
      border-width: 0;
      border-color: ${theme.colors.n900};
    }

    &:active,
    &:hover,
    &:hover:focus,
    &:active:focus {
      padding: ${calculatePadding({ theme, flat, secondary, ...otherProps })()};
    }

    &:active,
    &:hover,
    &:focus {
      color: ${theme.colors.n900};
    }

    &:active:focus,
    &:hover:focus {
      border-color: ${theme.colors.n900};
      border-width: 2px;
      box-shadow: none;
      padding: ${calculatePadding({ theme, flat, secondary, ...otherProps })(
        '2px'
      )};
    }
  `;

const sizeStyles = props => {
  const { size: buttonSize } = props;
  const padding = calculatePadding(props)();
  return css({
    label: `button--${buttonSize}`,
    padding
  });
};

const buttonLoadingStyles = ({ isLoading }) =>
  isLoading &&
  css`
    label: button--loading;
    pointer-events: none;
  `;

const ButtonElement = styled('button')`
  ${baseStyles};
  ${primaryStyles};
  ${sizeStyles};
  ${flatStyles};
  ${secondaryStyles};
  ${stretchStyles};
  ${buttonLoadingStyles};
`;

const LinkButtonElement = ButtonElement.withComponent('a');

const Button = ({ href, ...props }) =>
  href ? (
    <LinkButtonElement {...{ ...props, href }} />
  ) : (
    <ButtonElement {...props} />
  );

Button.propTypes = {
  /**
   * URL the Button should lead to. Causes the Button to render an <a> tag.
   */
  href: PropTypes.string,
  /**
   * Should the Button be disabled?
   */
  disabled: PropTypes.bool,
  /**
   * Button has a 'flat' variation, triggered with this prop.
   */
  flat: PropTypes.bool,
  /**
   * Renders a secondary button. Secondary buttons look the same for
   * primary (default) and flat buttons.
   */
  secondary: PropTypes.bool,
  /**
   * Renders a primary button using the brand color.
   */
  primary: PropTypes.bool,
  /**
   * Link target. Should only be passed, if href is passed, too.
   */
  target: PropTypes.string,
  /**
   * Size of the button. Use the Button's KILO, MEGA, or GIGA properties.
   */
  size: PropTypes.oneOf([Button.KILO, Button.MEGA, Button.GIGA]),
  /**
   * Trigger stretch (full width) styles on the component.
   */
  stretch: PropTypes.bool
};

Button.defaultProps = {
  disabled: false,
  flat: false,
  size: MEGA,
  target: null,
  href: null,
  primary: false,
  secondary: false,
  stretch: false
};

/**
 * @component
 */
export default Button;
