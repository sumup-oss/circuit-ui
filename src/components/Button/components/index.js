export { default as Button } from './Button';
export { default as LoadingIcon } from './LoadingIcon';
