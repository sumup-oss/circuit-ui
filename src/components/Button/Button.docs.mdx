import { Status, Props, Story } from '../../../.storybook/components';

import Button from './Button';
import ButtonGroup from '../ButtonGroup';
import LoadingButton from '../LoadingButton';

# Button

<Status.InReview />

A button allows us to highlight actions in the product and make them easily
perceived and achievable through just one click.

<Story id="components-button--base" />
<Props of={Button} />

## When to use it

Use buttons when users need to either navigate through a product or perform a specific task.

## Usage guidelines

- **Do** always label a button in a clear and understandable way
- **Do** always think about the priority of the action to be taken when choosing a type of button
- **Do** position buttons in a predictable and coherent way throughout the product.
- **Do** combine different button types if you have more than one button in a specific screen
- **Do** use a loading state to provide feedback for our user when the action revolves around saving or
  inputting information, especially when it could take some seconds.

## Content guidelines

- **Do** always start the label with an actionable verb accompanied by a noun to provide enough context (example: "Confirm order").
  The only exceptions are the following common actions: "Save"; "Cancel"; and "Close".
- **Do** keep the label to a maximum of 3 words
- **Do not** capitalize all words of the label, just the first word or brand/product words (such as Apple Pay, for example)

## Component variations

### Primary button

Used to highlight the most important action in a certain page. In most cases, there should be only one primary button visible
at a time.

<Story id="components-button--primary" />

### Default button

Used to highlight secondary actions to be takes, specially when there's already a primary action on the page.

<Story id="components-button--base" />

### Secondary button

Only use secondary buttons when paired to a primary button, placed in the left side of it. They are used to
signal opposite actions from the primary button, such as "Cancel order".

<Story id="components-button--secondary" />

### Plain button

Used when you want to highlight an action within a body text. It should always open in the same window unless
it links to an external source (such as a link to a complete study about card acceptance within a blog post).

<Story id="components-button--plain" />

### Button sizes

The button component also supports three different sizes: Mega, the best suitable for most cases; Kilo, used
when there is sizing constraints in the button position; and Giga, used for extreme cases when there's little
or no constraint on the button's placement.

<Story id="components-button--size" />

## Button group

Used when our user has two opposites actions to be taken in a certain step of a flow. It is generally used aligned to the right,
with a primary button for the expected action and a secondary button on its left.

<Story id="components-button-buttongroup--base" />
<Props of={ButtonGroup} />

- **Do** use the same verb tenses for both actions

## Loading button

When the button's action is inputting/saving information or it takes more than 3 seconds to take the user to the next step/page, you can indicate a loading state with `LoadingButton`, as well as visually notify the user of success or an error after loading. You should only use it on primary buttons.

<Story id="components-button-loadingbutton--success-animation" />
<Story id="components-button-loadingbutton--error-animation" />
<Story id="components-button-loadingbutton--no-exit-animation" />
<Props of={LoadingButton} />
