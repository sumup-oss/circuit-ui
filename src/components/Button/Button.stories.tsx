/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { action } from '@storybook/addon-actions';
import { ThumbUp } from '@sumup/icons';

import { Stack } from '../../../.storybook/components';
import ButtonGroup from '../ButtonGroup';
import LoadingButton from '../LoadingButton';
import IconButton from '../IconButton';
import CloseButton from '../CloseButton';

import { Button, ButtonProps } from './Button';
import docs from './Button.docs.mdx';

export default {
  title: 'Components/Button',
  component: Button,
  subcomponents: { LoadingButton, IconButton, CloseButton, ButtonGroup },
  parameters: {
    docs: { page: docs },
  },
  argTypes: {
    children: { control: 'text' },
  },
};

export const Base = (args: ButtonProps) => <Button {...args} />;

Base.args = {
  onClick: () => alert('Hello'),
  children: `Say hello`,
};

export const Variants = (args: ButtonProps) => (
  <Stack>
    <Button {...args} variant="primary">
      Primary
    </Button>
    <Button {...args} variant="secondary">
      Secondary
    </Button>
    <Button {...args} variant="tertiary">
      Tertiary
    </Button>
  </Stack>
);

export const Sizes = (args: ButtonProps) => (
  <Stack>
    <Button {...args} size="kilo">
      Kilo
    </Button>
    <Button {...args} size="mega">
      Mega
    </Button>
  </Stack>
);

export const WithIcon = (args: ButtonProps) => (
  <Button {...args} icon={ThumbUp}>
    Like
  </Button>
);

export const Tracking = (args: ButtonProps) => (
  <Button
    {...args}
    onClick={action('Button Click')}
    tracking={{
      label: 'track-button',
      customParameters: { key: 'value' },
    }}
  >
    Track me
  </Button>
);
