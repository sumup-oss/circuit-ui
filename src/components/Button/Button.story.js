/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Fragment } from 'react';
import { select, boolean, text } from '@storybook/addon-knobs/react';

import docs from './Button.docs.mdx';
import Button from './Button';

export default {
  title: 'Components/Button',
  component: Button,
  parameters: {
    docs: { page: docs },
    jest: ['Button']
  }
};

export const base = () => (
  <Button
    primary={boolean('Primary', false)}
    disabled={boolean('Disabled', false)}
    secondary={boolean('Secondary', false)}
    href={boolean('Link', false) ? '#' : undefined}
    target={boolean('Link', false) ? '_blank' : undefined}
    stretch={boolean('Stretched', false)}
    size={select('Size', [Button.KILO, Button.MEGA, Button.GIGA], Button.KILO)}
  >
    {text('Button Label', 'Button')}
  </Button>
);

export const primary = () => (
  <Button primary>{text('Button Label', 'Primary Button')}</Button>
);

export const secondary = () => (
  <Button secondary>{text('Button Label', 'Secondary Button')}</Button>
);

export const plain = () => (
  <Button plain href={boolean('href', false) ? 'http://sumup.com' : null}>
    {text('Button Label', 'Plain Button')}
  </Button>
);

export const size = () => (
  <Fragment>
    <Button size={Button.KILO}>Button kilo</Button>
    <Button size={Button.MEGA}>Button mega</Button>
    <Button size={Button.GIGA}>Button giga</Button>
  </Fragment>
);
