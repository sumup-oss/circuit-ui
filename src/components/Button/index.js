import Button from './ButtonContainer';

export default Button;
