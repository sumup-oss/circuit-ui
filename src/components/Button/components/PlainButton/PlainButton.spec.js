/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import Button from '../../Button';

describe('PlainButton', () => {
  /**
   * Style tests.
   */
  it('should render with default styles', () => {
    const actual = create(<Button plain>Link</Button>);
    expect(actual).toMatchSnapshot();
  });

  it('should render with href', () => {
    const actual = create(
      <Button plain href="example">
        Button
      </Button>
    );
    expect(actual).toMatchSnapshot();
  });

  it('should have primary styles', () => {
    const actual = create(
      <Button plain primary>
        Button
      </Button>
    );
    expect(actual).toMatchSnapshot();
  });

  it('should have kilo link styles', () => {
    const actual = create(
      <Button plain size={Button.KILO}>
        Button
      </Button>
    );
    expect(actual).toMatchSnapshot();
  });

  it('should have mega link styles', () => {
    const actual = create(
      <Button plain size={Button.MEGA}>
        Button
      </Button>
    );
    expect(actual).toMatchSnapshot();
  });

  it('should have giga link styles', () => {
    const actual = create(
      <Button plain size={Button.GIGA}>
        Button
      </Button>
    );
    expect(actual).toMatchSnapshot();
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines when used as anchor', async () => {
    const wrapper = renderToHtml(
      <Button plain href="http://accessibility.com">
        Link
      </Button>
    );
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });

  it('should meet accessibility guidelines when used as button', async () => {
    const wrapper = renderToHtml(
      <Button plain onClick={() => {}}>
        Link
      </Button>
    );
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
