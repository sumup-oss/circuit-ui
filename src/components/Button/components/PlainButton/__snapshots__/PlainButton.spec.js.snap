// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlainButton should have giga link styles 1`] = `
.circuit-0 {
  font-weight: 400;
  margin-bottom: 16px;
  font-size: 15px;
  line-height: 24px;
  margin-bottom: 0;
  box-shadow: none;
  border: none;
  background: none;
  font-weight: normal;
  padding: 0;
  color: #9DA7B1;
  cursor: pointer;
  outline: none;
}

@media (min-width:480px) {
  .circuit-0 {
    font-size: 18px;
    line-height: 28px;
  }
}

.circuit-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.circuit-0:active {
  color: #212933;
}

<button
  className="circuit-0 circuit-1"
  disabled={false}
  size="giga"
  target={null}
>
  Button
</button>
`;

exports[`PlainButton should have kilo link styles 1`] = `
.circuit-0 {
  font-weight: 400;
  margin-bottom: 16px;
  font-size: 13px;
  line-height: 20px;
  margin-bottom: 0;
  box-shadow: none;
  border: none;
  background: none;
  font-weight: normal;
  padding: 0;
  color: #9DA7B1;
  cursor: pointer;
  outline: none;
}

@media (min-width:480px) {
  .circuit-0 {
    font-size: 13px;
    line-height: 20px;
  }
}

.circuit-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.circuit-0:active {
  color: #212933;
}

<button
  className="circuit-0 circuit-1"
  disabled={false}
  size="kilo"
  target={null}
>
  Button
</button>
`;

exports[`PlainButton should have mega link styles 1`] = `
.circuit-0 {
  font-weight: 400;
  margin-bottom: 16px;
  font-size: 15px;
  line-height: 24px;
  margin-bottom: 0;
  box-shadow: none;
  border: none;
  background: none;
  font-weight: normal;
  padding: 0;
  color: #9DA7B1;
  cursor: pointer;
  outline: none;
}

@media (min-width:480px) {
  .circuit-0 {
    font-size: 15px;
    line-height: 24px;
  }
}

.circuit-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.circuit-0:active {
  color: #212933;
}

<button
  className="circuit-0 circuit-1"
  disabled={false}
  size="mega"
  target={null}
>
  Button
</button>
`;

exports[`PlainButton should have primary styles 1`] = `
.circuit-0 {
  font-weight: 400;
  margin-bottom: 16px;
  font-size: 15px;
  line-height: 24px;
  margin-bottom: 0;
  box-shadow: none;
  border: none;
  background: none;
  font-weight: normal;
  padding: 0;
  color: #9DA7B1;
  cursor: pointer;
  outline: none;
  color: #3388FF;
}

@media (min-width:480px) {
  .circuit-0 {
    font-size: 15px;
    line-height: 24px;
  }
}

.circuit-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.circuit-0:active {
  color: #212933;
}

.circuit-0:active {
  color: #003C8B;
}

<button
  className="circuit-0 circuit-1"
  disabled={false}
  size="mega"
  target={null}
>
  Button
</button>
`;

exports[`PlainButton should render with default styles 1`] = `
.circuit-0 {
  font-weight: 400;
  margin-bottom: 16px;
  font-size: 15px;
  line-height: 24px;
  margin-bottom: 0;
  box-shadow: none;
  border: none;
  background: none;
  font-weight: normal;
  padding: 0;
  color: #9DA7B1;
  cursor: pointer;
  outline: none;
}

@media (min-width:480px) {
  .circuit-0 {
    font-size: 15px;
    line-height: 24px;
  }
}

.circuit-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.circuit-0:active {
  color: #212933;
}

<button
  className="circuit-0 circuit-1"
  disabled={false}
  size="mega"
  target={null}
>
  Link
</button>
`;

exports[`PlainButton should render with href 1`] = `
.circuit-0 {
  font-weight: 400;
  margin-bottom: 16px;
  font-size: 15px;
  line-height: 24px;
  margin-bottom: 0;
  box-shadow: none;
  border: none;
  background: none;
  font-weight: normal;
  padding: 0;
  color: #9DA7B1;
  cursor: pointer;
  outline: none;
}

@media (min-width:480px) {
  .circuit-0 {
    font-size: 15px;
    line-height: 24px;
  }
}

.circuit-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.circuit-0:active {
  color: #212933;
}

<a
  className="circuit-0 circuit-1"
  disabled={false}
  href="example"
  size="mega"
  target={null}
>
  Button
</a>
`;
